### PR TITLE
alist: fix race condition while iterating

### PR DESCRIPTION
**Pull request by sebsura**
This pr fixes a race condition that happens when iterating over the same alist concurrently.  This is done by rippping out the internal iterator state and using extrernal ones.  Specifically we now use a simple pointer pair similar to std::vector.

#### Please check

- [X] Short description and the purpose of this PR is present _above this paragraph_
- [X] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Required backport PRs have been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

### CODE REVIEW

**Review by arogge — CHANGES_REQUESTED** [review]
I think we should preserve the ability to loop over a `NULL` `alist<T>*`. We just need free function templates for that which will also eliminate the need to dereference the `alist<T>*` before giving it to `for()`.

File: core/src/lib/alist.h [hunk 1]
```diff
@@ -64,8 +55,17 @@ enum
  *               array of pointers to inserted items
  */
 
+
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/lib/alist.h:58]: Great point!  I totally missed the fact that you can specify begin/end as free functions.

[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/lib/alist.h]: I replaced
```
if (x) for ( ... : *x )
```
with just
```
for ( ... : x )
```
but i left it as
```
for ( ... : *x )
```
in cases where we know that `x` is nonnull.

**Review by arogge — APPROVED** [review]
Great work! It is incredible how much simpler a lot of the code suddenly looks with just that simple change...

**Pull request merged**